src/stp/plays/defensive/DefendPass.cpp

```diff
@@ -53,29 +76,53 @@ void DefendPass::calculateInfoForRoles() noexcept {
     calculateInfoForDefenders();
     calculateInfoForKeeper();
     calculateInfoForHarassers();
+    calculateInfoForOffenders();
 }
 
 void DefendPass::calculateInfoForDefenders() noexcept {
     auto enemyRobots = world->getWorld()->getThem();
     auto enemyClosestToBall = world->getWorld()->getRobotClosestToBall(world::them);
 
-    enemyRobots.erase(std::remove_if(enemyRobots.begin(), enemyRobots.end(),
-                                     [&](const auto enemyRobot) -> bool { return enemyClosestToBall && enemyRobot->getId() == enemyClosestToBall.value()->getId(); }),
-                      enemyRobots.end());
+    stpInfos["defender_1"].setPositionToDefend(field.getOurTopGoalSide());
+    stpInfos["defender_1"].setBlockDistance(BlockDistance::PARTWAY);
+
+    stpInfos["defender_2"].setPositionToDefend(field.getOurBottomGoalSide());
+    stpInfos["defender_2"].setBlockDistance(BlockDistance::PARTWAY);
+
+    erase_if(enemyRobots, [&](const auto enemyRobot) -> bool { return enemyClosestToBall && enemyRobot->getId() == enemyClosestToBall.value()->getId(); });
+
+    auto enemyClosestToOurGoalOne = world->getWorld()->getRobotClosestToPoint(field.getOurGoalCenter(), enemyRobots);
+
+    erase_if(enemyRobots, [&](const auto enemyRobot) -> bool { return enemyClosestToOurGoalOne && enemyRobot->getId() == enemyClosestToOurGoalOne.value()->getId(); });
+
+    auto enemyClosestToOurGoalTwo = world->getWorld()->getRobotClosestToPoint(field.getOurGoalCenter(), enemyRobots);
+
+    erase_if(enemyRobots, [&](const auto enemyRobot) -> bool { return enemyClosestToOurGoalTwo && enemyRobot->getId() == enemyClosestToOurGoalTwo.value()->getId(); });
 
-    auto enemyClosestToOurGoal = world->getWorld()->getRobotClosestToPoint(field.getOurGoalCenter(), enemyRobots);
+    stpInfos["defender_helper_1"].setPositionToDefend(enemyClosestToOurGoalOne->get()->getPos());
+    stpInfos["defender_helper_1"].setBlockDistance(BlockDistance::HALFWAY);
 
-    enemyRobots.erase(std::remove_if(enemyRobots.begin(), enemyRobots.end(),
-                                     [&](const auto enemyRobot) -> bool { return enemyClosestToOurGoal && enemyRobot->getId() == enemyClosestToOurGoal.value()->getId(); }),
-                      enemyRobots.end());
+    stpInfos["defender_helper_2"].setPositionToDefend(enemyClosestToOurGoalTwo->get()->getPos());
+    stpInfos["defender_helper_2"].setBlockDistance(BlockDistance::HALFWAY);
 
-    auto remainingEnemy = world->getWorld()->getRobotClosestToPoint(field.getOurGoalCenter(), enemyRobots);
+    std::map<double, Vector2> enemyMap;
 
-    stpInfos["defender_1"].setPositionToDefend(enemyClosestToOurGoal->get()->getPos());
-    stpInfos["defender_1"].setBlockDistance(BlockDistance::HALFWAY);
+    for (auto enemy : enemyRobots) {
+        double score = FieldComputations::getPercentageOfGoalVisibleFromPoint(field, true, enemy->getPos(), world->getWorld().value(), enemy->getId(), false) *
```

Comment:
Although I think this way of distributing roles is quite clever, I think there is one issue with how it is done now. Midfielder_1 will not always go to the enemy with the highest goal visibility, which is fine, but as soon as midfielder_1 is standing in front of this enemy, its goal visibility will decrease significantly, which means that midfielder_1 will be assigned to a different robot (since a different enemy now has the highest goal visibility). This causes the robots to keep switching between enemies they are defending instead of actually standing in front of them to defend their pass.